File: src/Worker/content/NBB.Worker.FSharp/Messaging/SamplePublisherDecorator.fs
namespace NBB.Worker.Messaging

open Microsoft.Extensions.Logging
open NBB.Messaging.Abstractions
open System.Threading
open System

type SamplePublisherDecorator(inner: IMessageBusPublisher, logger: ILogger<SamplePublisherDecorator>) =
    interface IMessageBusPublisher with
        member _.PublishAsync<'T>(message: 'T, options : MessagingPublisherOptions, cancellationToken : CancellationToken) =
            let opts = if (isNull options) then MessagingPublisherOptions.Default else options

            let newCustomizer (outgoingEnvelope : MessagingEnvelope) =
                outgoingEnvelope.SetHeader("SampleHeader", "SampleValue")
                if (not(isNull opts.EnvelopeCustomizer)) then opts.EnvelopeCustomizer.Invoke(outgoingEnvelope)

            logger.LogDebug("Message {@Message} was sent.", message)
            inner.PublishAsync(message, MessagingPublisherOptions(EnvelopeCustomizer = Action<MessagingEnvelope>(newCustomizer), TopicName = opts.TopicName), cancellationToken)          

File: src/Worker/templates/NBB.Worker.CSharp/Messaging/SamplePublisherDecorator.cs
using Microsoft.Extensions.Logging;
using NBB.Messaging.Abstractions;
using System.Threading;
using System.Threading.Tasks;

namespace NBB.Worker.Messaging
{
    public class SamplePublisherDecorator : IMessageBusPublisher
    {
        private readonly IMessageBusPublisher _inner;
        private readonly ILogger<SamplePublisherDecorator> _logger;

        public SamplePublisherDecorator(IMessageBusPublisher inner, ILogger<SamplePublisherDecorator> logger)
        {
            _inner = inner;
            _logger = logger;
        }

        public Task PublishAsync<T>(T message, MessagingPublisherOptions options = null, CancellationToken cancellationToken = default)
        {
            options ??= MessagingPublisherOptions.Default;
            void NewCustomizer(MessagingEnvelope outgoingEnvelope)
            {
                outgoingEnvelope.SetHeader("SampleHeader", "SampleValue");
                options.EnvelopeCustomizer?.Invoke(outgoingEnvelope);
            }

            _logger.LogDebug("Message {@Message} was sent.", message);
            return _inner.PublishAsync(message, options with { EnvelopeCustomizer = NewCustomizer }, cancellationToken);
        }
    }
}